speed up download_tailored_datasets by adding option to parallel downloads

JasonFengGit opened the pull request:
# Pull Request

## Description

Trying to speed up download_tailored_datasets by adding an option to parallel downloads using `concurrent.futures`

Fixes https://github.com/openclimatefix/Satip/issues/144

With parallelization:
![image](https://github.com/openclimatefix/Satip/assets/40121574/53e0c454-3efd-44dd-8b0b-5c5e7102e929)



## How Has This Been Tested?

- [x] Yes
Tested with `test_data_tailor_parallel`, a new test I added in test_eumetsat.py. Parallelization reduces the test time from `18:10` to `07:41`.
![image](https://github.com/openclimatefix/Satip/assets/40121574/f4f0fc81-0338-4117-a45e-eb2faad00d56)


## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

jacobbieker reviewed (commented):
This looks good! If you can address the lint errors, we can merge it.

JasonFengGit commented:
Hi @jacobbieker, I've fixed the lint errors. However, I'm not sure if we should merge this into production before conducting more tests. I'm concerned about the potential memory overhead that might occur when running multiple downloads simultaneously.

I'll conduct some additional tests to see how it performs. Could you think of any scenarios where this could go wrong?

jacobbieker commented:
The data tailor only takes 3 jobs at a time, so potentially limiting it to a concurrency of 3 would be great, and keeps memory lower. Making it configurable is probably the best way forward for it.

JasonFengGit commented:
I see! Thanks!

JasonFengGit commented:
Hi @jacobbieker, I've added a concurrency param to limit the parallel downloads.

One concern I have is that a new test I added for parallel downloads takes about 8 minutes on my machine. This increases the total test time (which is already lengthy) from approximately 30 minutes to about 40 minutes. I am not sure if I should replace the original test_data_tailor, though. Thanks!

jacobbieker commented on satip/eumetsat.py:
```diff
@@ -432,27 +479,51 @@ def download_tailored_datasets(
+                            parent="DownloadManager"
+                        )
+        else:
+            for dataset_id in dataset_ids:
```
Actually, with thinking about it more, I would remove the non-parallel approach, and test. If the concurrency is defaulted to 1 instead, it still will test the futures work, and can be scaled up successfully, while cutting down on the test times.

JasonFengGit replied:
That makes sense! I'll implement that and test it out.

jacobbieker reviewed (commented):
Yeah, I agree the tests taking 40min is too long. Thinking about it more, I think its probably better to remove the non-parallel code, and the `parallel` option in the downloads. Instead only have the `concurrency` parameter, which should default to either 1 or 3, and then only have the parallel test, rather than both. That should speed up tests, and reduce the amount of code that will need to be maintained in the future.

JasonFengGit commented:
Hi @jacobbieker, I updated the code (remove the non-parallel approach and test, concurrency defaulted to 1) and tested on my machine. Thx!

JasonFengGit commented:
Hi @jacobbieker, speaking of testing, there is another reason for the long testing time: https://github.com/openclimatefix/Satip/issues/157#issuecomment-2023223214. Could you provide some suggestions for the implementation here? Thanks!

jacobbieker commented:
Hi, yeah, that sleep can probably be removed now, there can still be rate limiting, which that one is supposed to fix, but it seems to happen less often now. I would still leave the other sleeps that are around it which still would help mitigate hitting EUMETSAT's servers all at the same time

JasonFengGit commented:
I see! Could you review this pr or approve this https://github.com/openclimatefix/Satip/pull/247 so I can make a seperate PR? Thanks!

jacobbieker reviewed (approved):
This looks good! Thanks
